### PR TITLE
First cut code for loop deblocking filter

### PR DESCRIPTION
**Pull request by xiphmont**
No RDO as yet, but the deblocking filter and parameter infrastructure are
fully implemented, functional and tested.

Filter is currently turned on at a low-strength default; RDO is next.

### CODE REVIEW

File: src/deblock.rs [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+    if (po.x >= w) || (po.y >= h) {
+        return (0, 0, 0, 0, 0, 0)
+    }
+
+    // This little bit of weirdness is straight out of the spec;
+    // subsamped chroma uses odd mi row/col
+    bo.x |= xdec;
+    bo.y |= ydec;
+    let block = bc.at(&bo);
+    let block_adv = if pass == 0 { cmp::max(block.n4_w, 1<<xdec) } else { cmp::max(block.n4_h , 1<<ydec)};
```
Comment (lu-zero) [inline src/deblock.rs]: maybe this is more readable:

``` rust
let (block, block_adv) = if pass == 0  {
  (block.n4_w.max(1 << xdec), block.tx_w.max(1 << xdec))
} else {
  (block.n4_h.max(1 << xdec), block.tx_h.max(1 << xdec))
}

File: src/deblock.rs [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+        return (tx_adv, block_adv, 0, 0, 0, 0)
+    }
+
+    let mut level = deblock_level(fi, block, pli, pass) as u16;
+    if level == 0 { level = deblock_level(fi, prev_block, pli, pass) as u16; }
+    if level == 0 {
+        (tx_adv, block_adv, 0, 0, 0, 0)
+    } else {
+        
+        let tx_size = cmp::max(if pass == 0 { block.tx_w>>xdec } else { block.tx_h>>ydec } , 1);
```
Comment (lu-zero) [inline src/deblock.rs]: Same as above, if you like that style.

File: src/deblock.rs [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+    data[pitch*10] = (p2 + p1 + p0 + q0 + q1 + q2*2 + q3*2 + q4*2 + q5 + q6*4 + (1<<3) >> 4) as u16;
+    data[pitch*11] = (p1 + p0 + q0 + q1 + q2 + q3*2 + q4*2 + q5*2 + q6*5 + (1<<3) >> 4) as u16;
+    data[pitch*12] = (p0 + q0 + q1 + q2 + q3 + q4*2 + q5*2 + q6*7 + (1<<3) >> 4) as u16;
+}   
+
+fn deblock_edge4<'a>(slice: &'a mut PlaneMutSlice<'a>, pitch: usize, stride: usize,
+                 blimit: u16, limit: u16, thresh: u16, bd: usize) {
+    let mut s = 0;
+    let data = slice.as_mut_slice();
+    for _i in 0..4 {
```
Comment (lu-zero) [inline src/deblock.rs]:
``` rust
for p in data.chunks_mut(stride).take(4) {
  assert!(p.len() >= pitch * 3);
```

Should have mir and llvm more sure about what to do.

File: src/deblock.rs [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+        s += stride;
+    }
+}
+
+fn deblock_edge6<'a>(slice: &'a mut PlaneMutSlice<'a>, pitch: usize, stride: usize,
+                 blimit: u16, limit: u16, thresh: u16, bd: usize) {
+    let mut s = 0;
+    let flat = 1 << bd - 8;
+    let data = slice.as_mut_slice();
+    for _i in 0..4 {
```
Comment (lu-zero) [inline src/deblock.rs]:
``` rust
for p in data.chunks_mut(stride).take(4) {
  assert!(p.len() >= pitch * 5);
```

Should have mir and llvm more sure about what to do. (also see [slice patterns](https://rust-lang-nursery.github.io/edition-guide/rust-2018/slice-patterns.html#slice-patterns)).

File: src/deblock.rs [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+        s += stride;
+    }
+}
+
+fn deblock_edge8<'a>(slice: &'a mut PlaneMutSlice<'a>, pitch: usize, stride: usize,
+                 blimit: u16, limit: u16, thresh: u16, bd: usize) {
+    let mut s = 0;
+    let flat = 1 << bd - 8;
+    let data = slice.as_mut_slice();
+    for _i in 0..4 {
```
Comment (lu-zero) [inline src/deblock.rs]:
``` rust
for p in data.chunks_mut(stride).take(4) {
  assert!(p.len() >= pitch * 7);
```

Should have mir and llvm more sure about what to do. (same as above)

File: src/deblock.rs [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+        s += stride;
+    }
+}
+
+fn deblock_edge14<'a>(slice: &'a mut PlaneMutSlice<'a>, pitch: usize, stride: usize,
+                 blimit: u16, limit: u16, thresh: u16, bd: usize) {
+    let mut s = 0;
+    let flat = 1 << bd - 8;
+    let data = slice.as_mut_slice();
+    for _i in 0..4 {
```
Comment (lu-zero) [inline src/deblock.rs]: Same as above.

**Comment by barrbrain** [comment]
[AWCY results](https://beta.arewecompressedyet.com/?job=master-fd6d02d1d1b9f37d458105e3c70efede611da67c&job=rav1e-deblock-1%402018-09-01T07%3A53%3A44.021Z) at default speed:

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -0.2380 |  0.0815 |  0.1440 |  -0.2255 | -0.8731 | -0.9886 |    -0.2511 |

File: src/deblock.rs (hunk 2 above)
Comment (xiphmont) [reply src/deblock.rs]: yes, I should definitely do a readability pass, as well as add a bunch more commentary.  I got overexcited at it working :-)

File: src/deblock.rs (hunk 6 above)
Comment (xiphmont) [reply src/deblock.rs]: thanks!  I'll roll them in, with a few other changes I want to make, on monday.

File: src/deblock.rs (hunk 3 above)
Comment (xiphmont) [reply src/deblock.rs]: That can't work.  The loop is used for both row and column iteration depending on how its called, so stride can be equal to 1.

File: src/encoder.rs [hunk 7]
```diff
@@ -256,13 +255,11 @@ pub struct DeblockState {
 impl Default for DeblockState {
     fn default() -> Self {
         DeblockState {
-            levels: [0; PLANES+1],
+            levels: [16,16,8,8],
```
Comment (ycho) [inline src/encoder.rs]: Can you tell me what each entry of levels[] for? 
I read below lossless checking that both levels[0] > 0 and levels[1] > 0?

**Comment by ycho** [comment]
At default speed, objective-1-fast.

deblock-off@2018-09-03T00:39:08.348Z -> deblock-on@2018-09-03T00:38:21.303Z

   PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000
-2.8013 | -1.7360 | -1.5408 |  -1.7992 | -3.5996 | -3.3492 |    -2.5618

File: src/encoder.rs (hunk 7 above)
Comment (xiphmont) [reply src/encoder.rs]: pub levels: [u8; PLANES+1],  // Y vertical edges, Y horizontal, U, V

Comment (ycho) [reply src/encoder.rs]: Oh I see- thanks!

File: src/deblock.rs [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+    let ydec = p.cfg.ydec;
+    let po = bo.plane_offset(&p.cfg);
+        
+    // at or past edge of the frame?  Don't deblock, signal next loop.
+    if (po.x >= w) || (po.y >= h) {
+        return (0, 0, 0, 0, 0, 0)
+    }
+
+    // This little bit of weirdness is straight out of the spec;
+    // subsamped chroma uses odd mi row/col
```
Comment (ycho) [inline src/deblock.rs]: subsamp*l*ed

File: src/deblock.rs [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+    if !(block_edge || !block.skip || !prev_block.skip ||
+         block.ref_frames[0] <= INTRA_FRAME || prev_block.ref_frames[0] <= INTRA_FRAME) {
+        return (tx_adv, block_adv, 0, 0, 0, 0)
+    }
+
+    let mut level = deblock_level(fi, block, pli, pass) as u16;
+    if level == 0 { level = deblock_level(fi, prev_block, pli, pass) as u16; }
+    if level == 0 {
+        // When level == 0, the filter is a no-op even if it runs
+        (tx_adv, block_adv, 0, 0, 0, 0)
```
Comment (ycho) [inline src/deblock.rs]: Pls explain what '_adv' stanbd for?

Comment (ycho) [reply src/deblock.rs]: Aha... advance..!?

File: src/deblock.rs [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+            } else {
+                if tx_adv==0 {break};
+            }
+            tx += tx_adv;
+            if bx + block_adv == tx {bx = tx};
+        }
+    }
+}
+                 
+// Deblock horizontal edges in a single plane of a signle 64x64 superblock
```
Comment (ycho) [inline src/deblock.rs]: single

**Pull request merged**